11/umbraco-cms/extending/packages/good-practice-and-defaults.md

```diff
@@ -0,0 +1,207 @@
+
+```csharp
+// WRONG: Don't build paths manually
+var myPath = webrootPath + "\robots.txt";
+
+// CORRECT: Build the path using the correct separator for the current file system
+var myPath = Path.Combine(webrootPath, "robots.txt");
+```
+
+If you need to build a path manually, use `Path.DirectorySeparatorChar` instead to get the correct separator for the file system.
+
+## Settings
+
+Most packages will require some settings to be stored for the users to control in order to change the behavior of the package. Where you store these settings will depend a lot on the nature of the package.
+
+### Property Editors
+
+Property Editors should store their settings as part of their Data Type in Umbraco. This is the standard way property editor behavior is controlled while it is familiar to users and supported by deployment tools.
+
+### Do not save to appsettings.json
+
+You should not alter `appsettings.json` via code.
+
+Settings in ASP.NET Core are merged from a number of different locations at runtime. You cannot guarantee that `appsettings.json` is the location that a setting is read from and your users may not want certain settings in that file. You can read settings from the configuration, but you cannot assume they have come from `appsettings.json`.
+
+### Settings locations
+
+There are many options for where you might save your settings and a lot will depend on the nature of your package.
+
+Below you can find pros and cons for different places where you might save the settings for your package.
+
+#### Save to the database
+
+Settings can be saved to the database. Settings can be stored in the database using the Umbraco `IKeyValueservice`, and for more complex settings you can use a custom database table.
+
+- Pros:
+  - Settings will be accessible directly from the database, and not dependent on deployed files on disk.
+- Cons:
+  - Setup is required to create the database tables for the settings to live in.
+  - The settings will only be available to the specific instance of the site, and any settings will not be deployed between a local, development, or staging site.
+
+#### Save to disk
+
+You can choose to save the settings to disk. As an example, the settings can be saved in the `/config` folder at the root of the site.
+
+- Pros:
+  - Settings will be accessible to the site and can be included in deployments between sites.
+- Cons:
+  - You cannot guarantee that the folder or files will be present on a site or that they will be writable.
+  - Using your own config means your users cannot harness the power of the .NET Core options system and move settings to the environment or key stores. This means that sensitive information may end up on disk.
+
+#### Provide the users with appsettings.json snippets
+
+You could choose to provide your users with a snippet they can copy into their appsettings.json file. This will ensure that the settings are stored in the correct location.
```

Comment:
```suggestion
You could choose to provide your users with a snippet they can copy into their `appsettings.json` file. This will ensure that the settings are stored in the correct location.
```